### PR TITLE
Colored Folders

### LINKED ISSUE #319: Colored folders

### PR DESCRIPTION
**Pull request by KevinBoulongne**
Closes #319

### CODE REVIEW

File: app/src/main/res/layout/fragment_bottom_sheet_color_folder.xml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    <View
+        android:layout_width="26dp"
+        android:layout_height="@dimen/marginStandardVerySmall"
+        android:layout_gravity="center"
+        android:layout_marginVertical="@dimen/marginStandardMedium"
+        android:background="@drawable/background_bottom_sheet_anchor" />
+
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:layout_width="match_parent"
```
Comment (JorisBodin) [inline app/src/main/res/layout/fragment_bottom_sheet_color_folder.xml:34]: It's not correct
Show this: https://github.com/Infomaniak/android-kDrive/blob/master/app/src/main/res/layout/fragment_bottom_sheet_category_info_actions.xml#L39

File: app/src/main/res/layout/fragment_bottom_sheet_color_folder.xml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:orientation="vertical"
+    tools:context=".ui.bottomSheetDialogs.ColorFolderBottomSheetDialog">
+
+    <View
+        android:layout_width="26dp"
```
Comment (JorisBodin) [inline app/src/main/res/layout/fragment_bottom_sheet_color_folder.xml:27]: In a future PR, this view should be exported, as it does not have the same width everywhere.

File: app/src/main/java/com/infomaniak/drive/utils/Extensions.kt [hunk 3]
```diff
@@ -824,3 +837,10 @@ fun MaterialCardView.setCornersRadius(topCornerRadius: Float, bottomCornerRadius
         .setBottomRightCorner(CornerFamily.ROUNDED, bottomCornerRadius)
         .build()
 }
+
+fun File.isAllowedToBeColored(): Boolean {
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/utils/Extensions.kt:841]: Why an extension?
It should be directly in File

File: app/src/main/java/com/infomaniak/drive/utils/Extensions.kt [hunk 4]
```diff
@@ -278,24 +286,29 @@ fun View.setFileItem(file: File, isGrid: Boolean = false) {
 
     filePreview.scaleType = ImageView.ScaleType.CENTER
 
+    fun tintedDrawable(@IdRes iconId: Int, color: String? = null): Drawable? {
+        return ContextCompat.getDrawable(context, iconId)?.apply {
+            setTintList(ColorStateList.valueOf((color ?: file.getColor(context)).toColorInt()))
+        }
+    }
+
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/utils/Extensions.kt:294]: It should be in Utils.kt.

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewSliderFragment.kt [hunk 5]
```diff
@@ -155,6 +157,19 @@ class PreviewSliderFragment : Fragment(), FileInfoActionsView.OnItemClickListene
...
             bottomSheetFileInfos.refreshBottomSheetUi(currentPreviewFile)
         }
+
+        getBackNavigationResult<String>(ColorFolderBottomSheetDialog.COLOR_FOLDER_NAV_KEY) {
+            updateFolderColor(it)
+        }
+    }
+
+    private fun updateFolderColor(color: String) {
+        if (isResumed) {
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewSliderFragment.kt:167]: Why?

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileDetails/CreateOrEditCategoryFragment.kt [hunk 6]
```diff
@@ -126,7 +126,7 @@ class CreateOrEditCategoryFragment : Fragment() {
         categoriesRecyclerView.layoutManager = gridLayoutManager
     }
 
-    private fun getNumColorsColumns(minColumns: Int = 1, expectedItemSize: Int = 52): Int {
+    private fun getNumColorsColumns(minColumns: Int = 1, expectedItemSize: Int = 56): Int {
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/fileDetails/CreateOrEditCategoryFragment.kt:129]: Do we still need it?
What's wrong with AutoSpanGridLayoutManager?

**Review by JorisBodin — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/fragment_bottom_sheet_color_folder.xml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            android:id="@+id/coloredFolderText"
+            style="@style/Subtitle2"
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="@dimen/marginStandardMedium"
+            android:text="@string/buttonChangeFolderColor"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintStart_toEndOf="@id/coloredFolderIcon"
+            app:layout_constraintTop_toTopOf="parent" />
+    </androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (JorisBodin) [inline app/src/main/res/layout/fragment_bottom_sheet_color_folder.xml:59]: It's not correct
Show this: https://github.com/Infomaniak/android-kDrive/blob/master/app/src/main/res/layout/fragment_bottom_sheet_category_info_actions.xml#L39

File: app/src/main/java/com/infomaniak/drive/utils/Extensions.kt (hunk 4 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/drive/utils/Extensions.kt:294]: For now, this function is only used here.

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewSliderFragment.kt (hunk 5 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewSliderFragment.kt:167]: This shouldn't exist.

File: app/src/main/res/layout/fragment_bottom_sheet_color_folder.xml (hunk 2 above)
Comment (KevinBoulongne) [reply app/src/main/res/layout/fragment_bottom_sheet_color_folder.xml:27]: https://github.com/Infomaniak/android-kDrive/pull/478

File: app/src/main/java/com/infomaniak/drive/utils/Extensions.kt (hunk 4 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/drive/utils/Extensions.kt:294]: Done in another PR: https://github.com/Infomaniak/android-kDrive/pull/481

File: app/src/main/java/com/infomaniak/drive/ui/fileList/FileListFragment.kt [hunk 8]
```diff
@@ -606,18 +623,28 @@ open class FileListFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener {
...
+                BulkOperationType.COLOR_FOLDER -> {
+                    if (AccountUtils.getCurrentDrive()?.pack == Drive.DrivePack.FREE.value) {
+                        safeNavigate(R.id.colorFolderUpgradeBottomSheetDialog)
+                    } else {
+                        safeNavigate(FileListFragmentDirections.actionFileListToColorFolder(null))
+                    }
+                }
+                else -> {
+                    performBulkOperation(type)
                 }
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/FileListFragment.kt:647]: In one line plz :)

File: app/src/main/java/com/infomaniak/drive/ui/bottomSheetDialogs/ColorFolderBottomSheetDialog.kt [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import android.view.ViewGroup
+import androidx.navigation.fragment.navArgs
+import androidx.recyclerview.widget.RecyclerView
+import com.google.android.material.bottomsheet.BottomSheetDialogFragment
+import com.infomaniak.drive.R
+import com.infomaniak.drive.ui.bottomSheetDialogs.ColorFolderAdapter.Companion.COLORS
+import com.infomaniak.drive.utils.setBackNavigationResult
+import kotlinx.android.synthetic.main.fragment_bottom_sheet_color_folder.*
+
+open class ColorFolderBottomSheetDialog : BottomSheetDialogFragment() {
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/ui/bottomSheetDialogs/ColorFolderBottomSheetDialog.kt:32]: Why is open? 🤔

File: app/src/main/java/com/infomaniak/drive/ui/bottomSheetDialogs/ColorFolderBottomSheetDialog.kt [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        setAdapter()
+    }
+
+    private fun setAdapter() {
+        colorsAdapter = ColorFolderAdapter(
+            onColorSelected = { color: String -> setBackNavigationResult(COLOR_FOLDER_NAV_KEY, color) }
+        ).apply {
+            selectedPosition = COLORS.indexOfFirst { it == navigationArgs.color }
+            colorsRecyclerView.adapter = this
+            stateRestorationPolicy = RecyclerView.Adapter.StateRestorationPolicy.PREVENT_WHEN_EMPTY
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/ui/bottomSheetDialogs/ColorFolderBottomSheetDialog.kt:52]: Is it really necessary in this view?

File: app/src/main/java/com/infomaniak/drive/ui/bottomSheetDialogs/ColorFolderAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    var selectedPosition = -1
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
+        ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.view_color_round_button, parent, false))
+
+    override fun getItemCount(): Int = COLORS.size
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) = with(holder.itemView.colorButtonView) {
+        val color = ColorStateList.valueOf(COLORS[position].toColorInt())
+        val white = ContextCompat.getColorStateList(context, R.color.white)
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/ui/bottomSheetDialogs/ColorFolderAdapter.kt:43]: Why charge it in any case? It is not useful!

File: app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt [hunk 12]
```diff
@@ -170,6 +172,10 @@ object ApiRepository {
         return callApi(ApiRoutes.renameFile(file), POST, mapOf("name" to newName))
     }
 
+    fun updateFolderColor(file: File, color: String): ApiResponse<Boolean?> {
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt:175]: Can the boolean really be null?

File: app/src/main/java/com/infomaniak/drive/data/models/File.kt [hunk 13]
```diff
@@ -422,5 +432,13 @@ open class File(
...
             return java.io.File(mediaFolder, context.getString(R.string.EXPOSED_OFFLINE_DIR))
         }
+
+        fun File.getColor(context: Context): String {
+            return _color ?: String.format("#%06x", ContextCompat.getColor(context, R.color.secondaryText) and 0xffffff)
+        }
+
+        fun File.setColor(color: String) {
+            _color = color
+        }
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/data/models/File.kt:442]: Don't do extensions for this

Take inspiration from this: https://github.com/Infomaniak/android-kDrive/blob/master/app/src/main/java/com/infomaniak/drive/data/models/drive/Drive.kt#L75

**Review by JorisBodin — CHANGES_REQUESTED** [review]
You forgot the folder colour in `FileDetailsFragment`

File: app/src/main/java/com/infomaniak/drive/ui/bottomSheetDialogs/ColorFolderBottomSheetDialog.kt (hunk 9 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/drive/ui/bottomSheetDialogs/ColorFolderBottomSheetDialog.kt:32]: Mistake 😅

File: app/src/main/java/com/infomaniak/drive/ui/bottomSheetDialogs/ColorFolderBottomSheetDialog.kt (hunk 10 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/drive/ui/bottomSheetDialogs/ColorFolderBottomSheetDialog.kt]: Not anymore.

File: app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt (hunk 12 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt]: Oopsy!

File: app/src/main/java/com/infomaniak/drive/data/cache/FileMigration.kt [hunk 14]
```diff
@@ -98,6 +98,15 @@ class FileMigration : RealmMigration {
...
         }
+
+        // Migrated to version 3:
+        // - Added new field (Folder Color) in File table
+        if (oldVersionTemp == 2L) {
+            schema.get(File::class.java.simpleName)?.apply {
+                addField("_color", String::class.java)
+            }
+            oldVersionTemp++
+        }
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/data/cache/FileMigration.kt:109]: You haven't updated the migration :whip:

**Review by JorisBodin — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewSliderFragment.kt [hunk 15]
```diff
@@ -339,6 +339,10 @@ class PreviewSliderFragment : Fragment(), FileInfoActionsView.OnItemClickListene
         )
     }
 
+    override fun colorFolderClicked(color: String) {
+        // No-op
+    }
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewSliderFragment.kt:344]: No need to override this, if you set a default value from the `interface`

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileDetails/CreateOrEditCategoryFragment.kt [hunk 16]
```diff
@@ -120,7 +118,7 @@ class CreateOrEditCategoryFragment : Fragment() {
...
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/fileDetails/CreateOrEditCategoryFragment.kt:121]: 56 ?? Avoid magic value

**Review by sirambd — CHANGES_REQUESTED** [review]
Wait request changes

**Review by sirambd — APPROVED** [review]

**Pull request merged**